File: api/api.tork.challenge/api.tork.challenge.domain/Models/Book.cs
namespace api.tork.challenge.domain.Models
{
    public class Book
    {
        public int Id { get; set; }
        public string Title { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public int TotalCopies { get; set; }
        public int CopiesInUse { get; set; }
        public string Type { get; set; }
        public string Isbn { get; set; }
        public string Category { get; set; }

        public Book()
        {
            Validate();
        }

        private void Validate()
        {
            if (Id <= 0)
            { 
                throw new ArgumentNullException("Id is required");
            }

            if (string.IsNullOrEmpty(Title)) 
            { 
                throw new ArgumentNullException("Title is required");
            }

            if (string.IsNullOrEmpty(FirstName))
            {
                throw new ArgumentNullException("FirstName is required");
            }

            if (string.IsNullOrEmpty(LastName))
            {
                throw new ArgumentNullException("LastName is required");
            }

            if (TotalCopies < 0)
            {
                throw new ArgumentNullException("TotalCopies is invalid");
            }

            if (CopiesInUse < 0 || CopiesInUse > TotalCopies)
            {
                throw new ArgumentNullException("CopiesInUse is invalid");
            }

            if (string.IsNullOrEmpty(Type))
            {
                throw new ArgumentNullException("Type is required");
            }

            if (string.IsNullOrEmpty(Isbn))
            {
                throw new ArgumentNullException("Isbn is required");
            }

            if (string.IsNullOrEmpty(Category))
            {
                throw new ArgumentNullException("Category is required");
            }
        }
    }
}

File: app/src/domain/entity/book.ts
export default class Book
{
    private _id: number;
    private _title: string;
    private _firstName: string;
    private _lastName: string;
    private _totalCopies: number;
    private _copiesInUse: number;
    private _type: string;
    private _isbn: string;
    private _category: string;

    constructor(id: number, title: string, firstName: string, lastName: string,
        totalCopies: number, copiesInUse: number, type: string, isbn: string, category: string) 
    {
        this._id = id
        this._title = title
        this._firstName = firstName
        this._lastName = lastName
        this._totalCopies = totalCopies
        this._copiesInUse = copiesInUse
        this._type = type
        this._isbn = isbn
        this._category = category
    }

    get id(): number {
        return this._id;
    }

    get title(): string {
        return this._title;
    }

    get firstName(): string {
        return this._firstName;
    }

    get lastName(): string {
        return this._lastName;
    }

    get totalCopies(): number {
        return this._totalCopies;
    }

    get copiesInUse(): number {
        return this._copiesInUse;
    }

    get isbn(): string {
        return this._isbn;
    }

    get type(): string {
        return this._type;
    }

    get category(): string {
        return this._category;
    }
}